### PR TITLE
Visitor can see menu

### PR DESCRIPTION
**Pull request by jysmys**
## User Story
[PT link](https://www.pivotaltracker.com/story/show/172642005)

As a visitor,
In order to get an overview of the restaurant's products
I would like to be able to see a menu

## Changes proposed in this pull request:
Created view menu component to get products from the backend and render it in a table. Added Cypress test for the component.

## What I have learned working on this feature:
Render the table from array

## Packages/Gems added:
axios

###
image

### CODE REVIEW

File: src/components/ViewMenu.jsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  componentDidMount() {
+    this.getViewMenu();
+  }
+  async getViewMenu() {
+    let result = await getMenu();
+    this.setState({ menu: result.data.products });
+  }
+  render() {
+    let menu;
+    if (this.state.menu != null) {
```
Comment (erikbjoern) [inline src/components/ViewMenu.jsx]: Is an empty array null? You could do use the `.empty?` also, I think. But if the test is passing I guess `!= null` works fine!

File: cypress/integration/visitor_can_see_menu.feature.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    cy.route({
+      method: "GET",
+      url: "http://localhost:3001/api/v1/products",
+      response: "fixture:products.json",
+    });
+    cy.visit("http://localhost:3001/");
+    cy.get("button#add").should("be.visible");
+    cy.get("button#minus").should("be.visible");
+    cy.get("td#numberOfItems").should("be.visible");
+    cy.get("button#plus").should("be.visible");
```
Comment (erikbjoern) [inline cypress/integration/visitor_can_see_menu.feature.js]: Technically, the buttons aren't within scope for this first feature I think, but I don't know if they should be removed, now that they're there.

**Review by erikbjoern — COMMENTED** [review]
It looks good, you just did a bit too much, like we did 😁

File: cypress/integration/visitor_can_see_menu.feature.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  });
+
+  it("no items in the menu", () => {
+    cy.route({
+      method: "GET",
+      url: "http://localhost:3001/api/v1/products",
+      response: { data: {} },
+    });
+    cy.visit("http://localhost:3001/");
+    cy.contains("Nothing on the menu at the moment");
```
Comment (designerofthing) [inline cypress/integration/visitor_can_see_menu.feature.js]: Typo:  'Nothing **in** the menu at the moment'

**Review by designerofthing — COMMENTED** [review]
looks good, just the typo that I saw.  Go Turtle!

File: cypress/integration/visitor_can_see_menu.feature.js (hunk 2 above)
Comment (oliverochman) [reply cypress/integration/visitor_can_see_menu.feature.js]: please remove them, out of scope

We are testing the functionality for seeing a menu here, not seeing add to order buttons

File: cypress/fixtures/products.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "products": [
```
Comment (oliverochman) [inline cypress/fixtures/products.json:2]: since we are testing the visibility of a menu and not one single product, please add at least one more product here

File: cypress/integration/visitor_can_see_menu.feature.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      method: "GET",
+      url: "http://localhost:3001/api/v1/products",
+      response: "fixture:products.json",
+    });
+    cy.visit("http://localhost:3001/");
+    cy.get("button#add").should("be.visible");
+    cy.get("button#minus").should("be.visible");
+    cy.get("td#numberOfItems").should("be.visible");
+    cy.get("button#plus").should("be.visible");
+    cy.contains("pizza");
```
Comment (oliverochman) [inline cypress/integration/visitor_can_see_menu.feature.js]: what contains pizza?
please look at a specific element

File: src/App.jsx [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+import ViewMenu from "./components/ViewMenu";
+
+const App = () => {
+  return (
+    <>
+      <h1>Turtle - Food fast</h1>
+      <ViewMenu />
```
Comment (oliverochman) [inline src/App.jsx]: Why name it `ViewMenu` and not just `Menu`?

File: src/components/ViewMenu.jsx (hunk 1 above)
Comment (oliverochman) [reply src/components/ViewMenu.jsx]: Yeah put a debugger here and try this out. I do not think that an empty array is equal to null.

You could try `!= []` instead

File: src/components/ViewMenu.jsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      menu = this.state.menu.map((product) => {
+        return (
+          <Table.Row key={product.id}>
+            <Table.Cell>{product.title}</Table.Cell>
+            <Table.Cell>{product.description}</Table.Cell>
+            <Table.Cell>{product.price} $</Table.Cell>
+            <Table.Cell>
+              <button id="minus">-</button>
```
Comment (oliverochman) [inline src/components/ViewMenu.jsx]: These buttons are out of scope for this feature. Why have buttons that do nothing?

File: src/index.js [hunk 8]
```diff
@@ -1,8 +1,10 @@
...
-import App from './App';
-import * as serviceWorker from './serviceWorker';
+import React from "react";
+import ReactDOM from "react-dom";
+import App from "./App";
+import * as serviceWorker from "./serviceWorker";
+import axios from "axios";
 
-ReactDOM.render(<App />, document.getElementById('root'));
+axios.default.baseURL = "http://localhost:3002/api/v1";
```
Comment (oliverochman) [inline src/index.js]: are you running the rails app on localhost:3002?
why?

File: src/modules/menu.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import axios from "axios";
+
+const getMenu = async () => {
+  // let headers = {
```
Comment (oliverochman) [inline src/modules/menu.js]: remove commented out code

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: src/App.jsx (hunk 6 above)
Comment (jysmys) [reply src/App.jsx]: ok, we can live with that!

File: src/index.js (hunk 8 above)
Comment (jysmys) [reply src/index.js]: changed it to heroku path

File: src/components/ViewMenu.jsx (hunk 1 above)
Comment (jysmys) [reply src/components/ViewMenu.jsx]: Removed the sad path, and no check if the response from backend is empty

File: src/components/Menu.jsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  }
+  render() {
+    let menu;
+    menu = this.state.menu.map((product) => {
+      return (
+        <Table.Row key={product.id} id={product.id}>
+          <Table.Cell id={product.title}>{product.title}</Table.Cell>
+          <Table.Cell>{product.description}</Table.Cell>
+          <Table.Cell id="price">{product.price} $</Table.Cell>
+          <Table.Cell>
```
Comment (oliverochman) [inline src/components/Menu.jsx]: all of these extra buttons are still out of scope

File: src/components/Menu.jsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  }
+  async getViewMenu() {
+    let result = await getMenu();
+    this.setState({ menu: result.data.products });
+  }
+  render() {
+    let menu;
+    menu = this.state.menu.map((product) => {
+      return (
+        <Table.Row key={product.id} id={product.id}>
```
Comment (oliverochman) [inline src/components/Menu.jsx]: set the id to `id={"product-" + product.id}`

File: src/components/Menu.jsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import React from "react";
+import { Table } from "semantic-ui-react";
+import { getMenu } from "../modules/menu";
+import { Component } from "react";
+
+class Menu extends Component {
+  state = {
+    menu: [],
+    message: "",
```
Comment (oliverochman) [inline src/components/Menu.jsx]: why do you have a message state?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: cypress/integration/visitor_can_see_menu.feature.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    cy.route({
+      method: "GET",
+      url: "http://localhost:3000/api/v1/products",
+      response: "fixture:products.json",
+    });
+  });
+
+  it("sucessfully", () => {
+    cy.visit("http://localhost:3001/");
+    cy.get("#product-1").within(() => {
```
Comment (erikbjoern) [inline cypress/integration/visitor_can_see_menu.feature.js:13]: There should probably be assertions for more than one product, since it's supposed to show a list.

**Review by erikbjoern — COMMENTED** [review]
One comment, everything else looks good!
